rir/src/compiler/pir/instruction.h

```diff
@@ -853,51 +875,55 @@ class FLIE(Subassign2_2D, 5, Effect::None, EnvAccess::Leak) {
     Value* idx2() { return arg(3).val(); }
 };
 
-class FLIE(Extract1_1D, 3, Effect::None, EnvAccess::Leak) {
+class FLIE(Extract1_1D, 3, Effects::Any(), EnvAccess::Leak) {
   public:
     Extract1_1D(Value* vec, Value* idx, Value* env, unsigned srcIdx)
         : FixedLenInstructionWithEnvSlot(PirType::val(),
                                          {{PirType::val(), PirType::val()}},
                                          {{vec, idx}}, env, srcIdx) {}
+    void updateType() override final { maskEffectsOnNonObjects(); }
 };
 
-class FLIE(Extract2_1D, 3, Effect::None, EnvAccess::Leak) {
+class FLIE(Extract2_1D, 3, Effects::Any(), EnvAccess::Leak) {
   public:
     Extract2_1D(Value* vec, Value* idx, Value* env, unsigned srcIdx)
         : FixedLenInstructionWithEnvSlot(PirType::val().scalar(),
                                          {{PirType::val(), PirType::val()}},
                                          {{vec, idx}}, env, srcIdx) {}
+    void updateType() override final { maskEffectsOnNonObjects(); }
 };
 
-class FLIE(Extract1_2D, 4, Effect::None, EnvAccess::Leak) {
+class FLIE(Extract1_2D, 4, Effects::Any(), EnvAccess::Leak) {
   public:
     Extract1_2D(Value* vec, Value* idx1, Value* idx2, Value* env,
                 unsigned srcIdx)
         : FixedLenInstructionWithEnvSlot(
               PirType::val(),
               {{PirType::val(), PirType::val(), PirType::val()}},
               {{vec, idx1, idx2}}, env, srcIdx) {}
+    void updateType() override final { maskEffectsOnNonObjects(); }
 };
 
-class FLIE(Extract2_2D, 4, Effect::None, EnvAccess::Leak) {
+class FLIE(Extract2_2D, 4, Effects::Any(), EnvAccess::Leak) {
   public:
     Extract2_2D(Value* vec, Value* idx1, Value* idx2, Value* env,
                 unsigned srcIdx)
         : FixedLenInstructionWithEnvSlot(
               PirType::val().scalar(),
               {{PirType::val(), PirType::val(), PirType::val()}},
               {{vec, idx1, idx2}}, env, srcIdx) {}
+    void updateType() override final { maskEffectsOnNonObjects(); }
```

Comment:
Isn't this more like a resetType?

Reply:
why reset? in general this is the api that we call after some optimization passes. it means: given your new input state, can you give a more precise output type. this is not reseting the output type. it is updating it. So I don't understand in what sense that would be resetting (which means setting something to the value it initially was).

Reply:
I meant the `maskEffectsOnNonObjects()` method. 
The updateType on arithmetic ops is only ensuring that it is still safe to "hide", and otherwise it "reset" those flags to the original state. 
The name `maskEffectsOnNonObjects` sounded a little low-level to me, we are reseting those flags.